sql/src/main/java/io/crate/expression/scalar/cast/CastFunction.java

```diff
@@ -28,26 +28,119 @@
 import io.crate.expression.symbol.Function;
 import io.crate.expression.symbol.Literal;
 import io.crate.expression.symbol.Symbol;
-import io.crate.metadata.BaseFunctionResolver;
 import io.crate.metadata.FunctionIdent;
-import io.crate.metadata.FunctionImplementation;
 import io.crate.metadata.FunctionInfo;
+import io.crate.metadata.FunctionName;
 import io.crate.metadata.Scalar;
 import io.crate.metadata.TransactionContext;
-import io.crate.metadata.functions.params.FuncParams;
-import io.crate.metadata.functions.params.Param;
+import io.crate.metadata.functions.Signature;
 import io.crate.types.DataType;
-import io.crate.types.UndefinedType;
 
-import java.util.List;
 import java.util.Map;
 import java.util.function.BiFunction;
 
 import static io.crate.expression.scalar.cast.CastFunctionResolver.CAST_SIGNATURES;
 import static io.crate.expression.scalar.cast.CastFunctionResolver.TRY_CAST_PREFIX;
+import static io.crate.metadata.functions.TypeVariableConstraint.typeVariable;
+import static io.crate.types.TypeSignature.parseTypeSignature;
 
 public class CastFunction extends Scalar<Object, Object> {
 
+    public static void register(ScalarFunctionModule module) {
+        // We still maintain the cast function to type mapping to stay
+        // bwc by keeping the old `to_<type>` and `try_<type>` function signatures.
+        //
+        // We can drop the per type cast function already in 4.2. This change
+        // would require handling the metadata for the places where the old
+        // signature of the cast function were used, e.g. generated columns.
+        for (Map.Entry<String, DataType> function : CAST_SIGNATURES.entrySet()) {
+            module.register(
+                Signature.builder()
+                    .name(new FunctionName(null, function.getKey()))
+                    .kind(FunctionInfo.Type.SCALAR)
+                    .typeVariableConstraints(typeVariable("E"), typeVariable("V"))
+                    .argumentTypes(parseTypeSignature("E"), parseTypeSignature("V"))
```

Comment:
I have not changed the second argument, such is I restrict it `function.getValue().getTypeSignature()` then we won't much a function that takes as the second argument an object type with inner types. the generic `V` for the second argument works in this case. I also tried to drop coercion and explicitly provide the type info for the second argument, but then it loses the inner type info.

Reply:
We could  (and maybe must) add a special case for targetType = ObjectType with a more concrete signature of `parseTypeSignature("object(text, V)")` (including a variableArityGroup of `text, V`).
Currently all signature are the same and it will  always select the first one (with a probably wrong return type).
(This is btw. currently wrong, it should fail if multiple functions match, I'll provide a fix soon).
Loosing the innerType should not happen with your fix to `createType()` + `getTypeSignature()`.

Reply:
it is not possible to parse `parseTypeSignature("object(text, V)")` and call create type on it later on due to the assertion in the `createType` method. Also, there are probably also other cases, like `array(object)`, etc.

Hm but all function signatures are different, coz they encode the return type name into their names. For object type matching `parseTypeSignature("V")` works fine, and inner types are still there.

With this signature matching for two arguments here, it feels that we are trying to provide a correct solution on top of the existing workaround for the return type in the CastFunctionResolver.

Reply:
> it is not possible to parse parseTypeSignature("object(text, V)") and call create type on it later on due to the assertion in the createType method

Right but this should (and currently will) never happen. `createType()` should only be called on bound argument types which won't be the defined type signatures.

> Hm but all function signatures are different, coz they encode the return type name into their names.

Ah right, missed the different function names.
As we've discussed this looks like we should get rid of these different function names eventually as all will end up in the same signature + implementation. But yes, lets do this later.